### PR TITLE
[array API] add device property & to_device method

### PR DESCRIPTION
**Pull request by jakevdp**
Part of #21088 and #18353

As discussed earlier, `arr.device` will return the device for single-device arrays, and the sharding for non-trivially sharded arrays. This lets `arr.device` be passed to `device_put`, as well as to other Array API arguments that accept `Device | Sharding`.

### CODE REVIEW

File: jax/_src/array.py [hunk 1]
```diff
@@ -254,6 +254,13 @@ def size(self):
   def sharding(self):
     return self._sharding
 
+  @property
+  def device(self):
```
Comment (yashk2810) [inline jax/_src/array.py:258]: Why can't we return a Sharding always?

Comment (yashk2810) [reply jax/_src/array.py:258]: Cool, sounds good.

File: jax/_src/earray.py [hunk 2]
```diff
@@ -83,6 +83,13 @@ def sharding(self):
     phys_sharding = self._data.sharding
     return sharding_impls.logical_sharding(self.aval, phys_sharding)
 
+  @property
+  def device(self):
```
[the remark this answers is not in the document]
Comment (jakevdp) [reply jax/_src/earray.py:87]: That would return the physical sharding, I think we want to return the logical sharding.

File: jax/_src/numpy/array_methods.py [hunk 3]
```diff
@@ -67,6 +68,12 @@ def _astype(arr: ArrayLike, dtype: DTypeLike, copy: bool = False, device: xc.Dev
   """
   return lax_numpy.astype(arr, dtype, copy=copy, device=device)
 
+def _to_device(arr: ArrayLike, device: xc.Device | Sharding | None, *,
```
[the remark this answers is not in the document]
Comment (jakevdp) [reply jax/_src/numpy/array_methods.py]: Done

File: jax/_src/earray.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yashk2810) [reply jax/_src/earray.py:87]: Ahh, right. Cool.

**Review by yashk2810 — APPROVED** [review]

**Pull request merged**